File: FrontEnd/src/services/bookingService.js
import API from "../api/axiosConfig";

const createBooking = async (bookingData) => {
    const response = await API.post("/bookings/create", bookingData);
    return response.data;
};

const confirmBooking = async (bookingId) => {
    const response = await API.put(`/bookings/${bookingId}/confirm`);
    return response.data;
};

const cancelBooking = async (bookingId, reason) => {
    const url = reason 
        ? `/bookings/${bookingId}/cancel?reason=${encodeURIComponent(reason)}` 
        : `/bookings/${bookingId}/cancel`;
    const response = await API.put(url);
    return response.data;
};

const getBookingById = async (bookingId) => {
    const response = await API.get(`/bookings/${bookingId}`);
    return response.data;
};

const getBookingsByUser = async (userId) => {
    const response = await API.get(`/bookings/user/${userId}`);
    return response.data;
};

const getBookingsByBus = async (busId) => {
    const response = await API.get(`/bookings/bus/${busId}`);
    return response.data;
};

const calculateFare = async (busId, numberOfSeats) => {
    const response = await API.get(`/bookings/fare?busId=${busId}&numberOfSeats=${numberOfSeats}`);
    return response.data;
};

export default {
    createBooking,
    confirmBooking,
    cancelBooking,
    getBookingById,
    getBookingsByUser,
    getBookingsByBus,
    calculateFare,
};

File: FastX_Backend/src/main/java/com/hexaware/fastx/service/BookingService.java
package com.hexaware.fastx.service;

import java.util.List;

import com.hexaware.fastx.dto.BookingDto;

public interface BookingService {
    BookingDto createBooking(BookingDto bookingDto);
    BookingDto confirmBooking(Long bookingId);
    BookingDto cancelBooking(Long bookingId);
    BookingDto cancelBooking(Long bookingId, String reason);
    BookingDto getBookingById(Long bookingId);
    List<BookingDto> getBookingsByUser(Long userId);
    List<BookingDto> getBookingsByBus(Long busId);
    Double calculateFare(Long busId, int numberOfSeats);

    /**
     * Verifies that the given booking belongs to the specified user.
     * Throws UnauthorizedException if the booking belongs to a different user.
     * Used by controllers to enforce USER ownership before confirm/cancel.
     */
    void verifyBookingOwnership(Long bookingId, Long userId);

    /**
     * Verifies that the given bus belongs to the specified operator.
     * Throws UnauthorizedException if the bus belongs to a different operator.
     */
    void verifyBusOperatorOwnership(Long busId, Long operatorId);
}
